### ISSUE TITLE
syntax: inline backtick command substitutions can fail to parse or break programs

### ISSUE THREAD

**Issue by aksh1618** [mvdan/sh#342]
I am using command substitution for [comments in multi line commands](https://stackoverflow.com/a/12797512/6346531):

```bash
$xdotool_command \
    windowactivate --sync $spotify_window_id `# activate spotify window` \
    mousemove $x $y                          `# move mouse to favorite button` \
    click 1                                  `# click left mouse button` \
    windowactivate $active_window_id         `# activate current window`
```

But it gets formatted to this:

```bash
$xdotool_command \
    windowactivate --sync $spotify_window_id $(# activate spotify window` \
        mousemove $x $y
    ) # move mouse to favorite button` \
click 1 $(# click left mouse button` \
    windowactivate $active_window_id
) # activate current window`
```

**Comment by mvdan** [comment]
Wow, I had never seen that way to have inline comments within multiline commands. It's perhaps the only time I've seen them being useful compared to `$()`, though I'm not sure how I feel about the hack.

The first issue that comes to mind is that backticks are pretty much deprecated. See https://unix.stackexchange.com/questions/126927/have-backticks-i-e-cmd-in-sh-shells-been-deprecated.

`shfmt` is an opinionated tool, so backticks are removed automatically. There currently is no way to turn this off, and noone has complained for over three years:

```
$ shfmt <<<'echo `foo`'
echo $(foo)
```

So on one hand, I don't think the software should encourage the use of backticks. And on the other, I'm not even sure how we would implement this if we wanted to. Adding a flag to allow backticks would seem like endorsement for their use. Trying to guess which backticks are actually just inline comments would very likely be a complex and buggy heuristic.

To summarise - while I understand that inline comments in this scenario are a pain, I don't think it's a good idea to support the mentioned hack in `shfmt`. There are plenty of less sintactically hacky workarounds, such as placing comments before the command, or if the command would span many lines, building a command's arguments with an array:

```
# bash
args=(
    arg1 # arg1 comment
    arg2 # arg2 comment
    ...
)
cmd "${args[@]}"

# posix
args=""
args+=" arg1" # arg1 comment
args+=" arg2" # arg2 comment
...
cmd "$args"
```

Of course these workarounds add a bit of an overhead, but so do command substitutions in general. And I presume readability and portability are much more important than performance for a shell script. Otherwise one would use a compiled language :)

**Comment by aksh1618** [comment]
> Trying to guess which backticks are actually just inline comments would very likely be a complex and buggy heuristic.

Wouldn't it just involve checking for a backtick immediately followed by a `#` and a subsequent backtick? Just guessing, I'm not well versed in this. Also, shouldn't the current behavior be considered a bug? It does convert working code to failing code, or simply fails in case of odd number of lines with the comments, as it doesn't see the ending backticks.

> building a command's arguments with an array

Hadn't seen this before, thanks! Kinda torn between the two now :sweat_smile:

**Comment by mvdan** [comment]
> Wouldn't it just involve checking for a backtick immediately followed by a # and a subsequent backtick? 

In the simplest form, yes. But there's a few edge cases I can think of:

* Multiple of these on a single line
* Newlines or spaces around the comment
* One of these hacky comments right before a normal comment
* Nested hacky comments, or hacky comments within other backquotes

And of course, we'd likely find many more in the wild if the feature was implemented :) I personally think it's not worth the hassle - neither for the developer nor for the tool.

> Also, shouldn't the current behavior be considered a bug? It does convert working code to failing code, or simply fails in case of odd number of lines with the comments, as it doesn't see the ending backticks.

Technically, yes. In practice, backticks are insane to parse correctly. They do a bunch of pre-processing to the input bytes, which is easy for Bash to do as it parses entire files at once in memory. In our case, the parser uses a stream of bytes, so these kinds of pre-processing steps are hard to get right.

Think about it - if you give the shell `` foo `#comment` bar ``, it goes "ah, I see backticks, so those bytes will be parsed as if EOF was at the closing backtick!". All the notion of a comment being ended by a newline or EOF in the file goes out the window, as suddenly the shell decides to parse what's in between the backticks as if it were an entirely new file. This is one of the reasons why `$()` are saner and recommended.

I'll leave this issue open to investigate how hard it would be to fix this edge case of backtick parsing. For example, last year I fixed parsing nested backticks, as that requires some wacky code to handle the extra backslashes.

That piece was about the parser. The printer will likely keep converting bacticks to `$()`. In the case where that effectively breaks the program, I'm not sure what to do. We have multiple options:

* Keep the backticks in that edge case (really dislike this option)
* Keep breaking programs (what we currently do; not ideal)
* Move these comments elsewhere, e.g. before or after the command in question

I might give that last option a try. Seems somewhat better than straight up breaking the program.

**Comment by kaey** [comment]
Example in OP breaks and has to be rewritten, when some arguments need to be added based on some condition. Here is part from my qemu script for example
```
command=qemu-system-x86_64
arg() {
        command="$command $@"
}

arg -device usb-tablet
arg -nic bridge,br=bridge0,model=$drv_net,mac=$mac

if [ -z "$nodisplay" ]; then
        arg -daemonize
        arg -display gtk
        arg -serial vc
        arg -monitor vc
        arg -vga $vga
else
        arg -nographic
        arg -chardev stdio,mux=on,signal=off,id=char0
        arg -serial chardev:char0
        arg -mon chardev=char0,mode=readline
        arg -vga std
fi

if [ -z "$nouefi" ]; then
        arg -drive if=pflash,format=raw,readonly,file=/usr/share/ovmf/x64/OVMF_CODE.fd
        arg -drive if=pflash,format=raw,file=$uefi_vars
fi

eval $command $extra_args
```

Supporting behavior which does not work in usual cases may not be worth the cost of complicating parser

**Comment by mvdan** [comment]
I'm inclined to not do anything here. If someone has code that relies on backticks and "inline comments" like this, I'm sorry, but they're not supported. We remove backticks anywaya, and I don't want to change that rule.

If you really must need inline comments in shell, either look at a different approach, or use a different language.

**Comment by mvdan** [comment]
Ended up implementing this in the PR above; reviews and testing welcome :)
